File: crates/outlines/testdata/python/complex.py
class Reader:
    def read(self, key: str) -> str:
        return key.upper()

    def close(self) -> None:
        return None


class CachedReader(Reader):
    def __init__(self) -> None:
        self.cache: dict[str, str] = {}

    def read(self, key: str) -> str:
        if key not in self.cache:
            self.cache[key] = super().read(key)
        return self.cache[key]


class Service:
    class Metrics:
        def __init__(self) -> None:
            self.calls = 0

    def __init__(self, reader: Reader) -> None:
        self.reader = reader
        self.metrics = self.Metrics()

    def run(self, key: str) -> str:
        self.metrics.calls += 1
        return self.reader.read(key)

    def close(self) -> None:
        self.reader.close()


def bootstrap(reader: Reader) -> Service:
    service = Service(reader)
    service.run("startup")
    return service

File: crates/outlines/testdata/typescript/complex.ts
export type EventMap = {
  ready: Date
  error: Error
}

export interface Reader {
  read(key: string): string
  close(): void
}

export enum Status {
  Idle,
  Busy,
  Closed,
}

export class Service {
  static cache = new Map<string, string>()

  constructor(
    private reader: Reader,
    private status: Status = Status.Idle,
  ) {}

  run(key: string) {
    this.status = Status.Busy
    const value = this.reader.read(key)
    Service.cache.set(key, value)
    return value
  }

  close() {
    this.reader.close()
    this.status = Status.Closed
  }

  static from(reader: Reader) {
    return new Service(reader)
  }
}

export function bootstrap(reader: Reader) {
  const service = Service.from(reader)
  service.run("startup")
  return service
}
